File: code/community/1301691162435784711/integrating-stt-tts-deepgram.py
import os
import asyncio
import websockets

async def connect_stt():
    url = "wss://api.deepgram.com/v1/listen"
    api_key = os.environ['DEEPGRAM_API_KEY']
    headers = {"Authorization": f"Token {api_key}"}

    async with websockets.connect(url, extra_headers=headers) as websocket:
        await websocket.send("{\"content-type\": \"audio/wav\", \"interim_results\": true}")
        while True:
            response = await websocket.recv()
            print(f"STT Received: {response}")

async def connect_tts():
    url = "wss://api.deepgram.com/v1/speak"
    api_key = os.environ['DEEPGRAM_API_KEY']
    headers = {"Authorization": f"Token {api_key}"}

    async with websockets.connect(url, extra_headers=headers) as websocket:
        tts_text = "Hello, this is a test message for TTS."
        await websocket.send(tts_text)
        response = await websocket.recv()
        print(f"TTS Received: {response}")

async def main():
    await asyncio.gather(
        connect_stt(),
        connect_tts()
    )

if __name__ == '__main__':
    asyncio.run(main())

File: code/community/1301691162435784711/integrating-stt-tts-deepgram.rs
use std::env;
use std::error::Error;
use websocket::{ClientBuilder, Message};
use std::thread;
use std::io::Write;

fn main() -> Result<(), Box<dyn Error>> {
    let stt_url = "wss://api.deepgram.com/v1/listen";
    let tts_url = "wss://api.deepgram.com/v1/speak";
    let api_key = env::var("DEEPGRAM_API_KEY")?;

    // Connect to the STT WebSocket
    let stt_client = ClientBuilder::new(stt_url)?
        .add_protocol("rust-websocket")
        .connect_secure(Some(websocket::sync::NativeTlsSettings::default()))?;
    
    stt_client.send_message(&Message::text("{\"content-type\": \"audio/wav\", \"interim_results\": true}"))?;

    // Keep the connection open while processing
    let (mut receiver, mut sender) = stt_client.split()?;

    thread::spawn(move || {
        for message in receiver.incoming_messages() {
            match message {
                Ok(message) => match message {
                    Message::Text(txt) => println!("Received: {}", txt),
                    _ => {}
                },
                Err(e) => eprintln!("Error: {:#?}", e),
            }
        }
    });

    // Example TTS management (assuming TTS state is managed similarly)
    let tts_client = ClientBuilder::new(tts_url)?
        .add_protocol("rust-websocket")
        .connect_secure(Some(websocket::sync::NativeTlsSettings::default()))?;
    
    // For TTS, send a message to play text
    let tts_text = "Hello, this is a test of Text-to-Speech API.";
    tts_client.send_message(&Message::text(tts_text))?;

    std::io::stdout().flush()?;

    Ok(())
}